### PR TITLE
Add `sfBuffer` wrapper and implement `sf::Image::saveToMemory`

### PR DESCRIPTION
**Pull request by kimci86**
Alternative to #195 using a C wrapper over an `std::vector`.

### CODE REVIEW

File: src/SFML/Graphics/Buffer.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    CSFML_CHECK_RETURN(buffer, 0);
+    return buffer->vector.size();
+}
+
+
+////////////////////////////////////////////////////////////
+CSFML_GRAPHICS_API const sfUint8* sfBuffer_getData(const sfBuffer* buffer)
+{
+    CSFML_CHECK_RETURN(buffer, NULL);
+    return buffer->vector.data();
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Buffer.cpp:59]: Do we want to `assert(buffer->vector.size() > 1)`? The standard specifies that if the vector is empty then `.data()` may or may not return a null pointer. I suppose it's not necessarily a bug to call it under those circumstances but it's worth consideration.

File: src/SFML/Graphics/Image.cpp [hunk 2]
```diff
@@ -131,6 +133,14 @@ sfBool sfImage_saveToFile(const sfImage* image, const char* filename)
 }
 
 
+////////////////////////////////////////////////////////////
+sfBool sfImage_saveToMemory(const sfImage* image, sfBuffer* output, const char* format)
+{
+    CSFML_CHECK_RETURN(output, sfFalse);
+    CSFML_CALL_RETURN(image, saveToMemory(output->vector, format), sfFalse);
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/Image.cpp:140]: `format` gets converted to `std::string` and it's UB to construct a string from a null pointer. Should we add a `CSFML_CHECK_RETURN` call for `format` being null?

**Review by ChrisThrasher — APPROVED** [review]
I appreciate you putting this together with all the right docs and everything.

File: src/SFML/Graphics/Buffer.cpp (hunk 1 above)
Comment (kimci86) [reply src/SFML/Graphics/Buffer.cpp:59]: The valid range for any vector being [data(), data() + size()), the valid range for an empty vector is empty. I don't think we need a special case. It is up to the user not to dereference values out of the valid range.

File: src/SFML/Graphics/Image.cpp (hunk 2 above)
Comment (kimci86) [reply src/SFML/Graphics/Image.cpp:140]: Nice catch, will do.

Comment (ChrisThrasher) [reply src/SFML/Graphics/Image.cpp:140]: That's fine by me

File: include/SFML/Graphics/Buffer.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#include <stddef.h>
+
+
+////////////////////////////////////////////////////////////
+/// \brief Create an empty buffer
+///
+/// \return A new sfBuffer object, or NULL if it failed
+///
+////////////////////////////////////////////////////////////
+CSFML_GRAPHICS_API sfBuffer* sfBuffer_create();
```
Comment (kimci86) [inline include/SFML/Graphics/Buffer.h:42]: I forgot the `void` parameter, will fix.

File: src/SFML/Graphics/Buffer.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/Buffer.cpp:59]: `data()` is C++11, I guess we should keep CSFML 2.6 aligned with the C++ versioning if possible

File: src/SFML/Graphics/BufferStruct.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#include <SFML/Config.hpp>
+#include <vector>
+
+
+////////////////////////////////////////////////////////////
+// Internal structure of sfBuffer
+////////////////////////////////////////////////////////////
+struct sfBuffer
+{
+    std::vector<sf::Uint8> vector;
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/BufferStruct.h:40]: I would call the member after the container type, maybe just `buffer` instead?

File: include/SFML/Graphics/Buffer.h
Comment (eXpl0it3r) [inline include/SFML/Graphics/Buffer.h:1]: Kind of wondering whether this should live in the System part, but I guess as long as we just use it in the Graphics module, lets keep it here.

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]

File: src/SFML/Graphics/Buffer.cpp (hunk 4 above)
Comment (kimci86) [reply src/SFML/Graphics/Buffer.cpp]: Well spotted! Fixed.

File: src/SFML/Graphics/BufferStruct.h (hunk 5 above)
Comment (kimci86) [reply src/SFML/Graphics/BufferStruct.h]: Ok, done

**Comment by kimci86** [comment]
I used the following code to test the API.
It made me notice the `#include <SFML/Graphics/Buffer.h>` was missing from `SFML/Graphics.h` file and fixed it.

```c
#include <SFML/Graphics.h>
#include <stdio.h>

int main(void)
{
    const int WIDTH = 256;
    sfImage* image = sfImage_createFromColor(WIDTH, 1, sfMagenta);
    for (int x = 0; x < WIDTH; ++x)
        sfImage_setPixel(image, x, 0, sfColor_fromRGB(x, 255, x));

    sfBuffer* buffer = sfBuffer_create();
    sfBool success = sfImage_saveToMemory(image, buffer, "png");
    printf("Succes: %d\n", success);

    sfImage_destroy(image);

    FILE* f = fopen("out.png", "wb");
    fwrite(sfBuffer_getData(buffer), 1, sfBuffer_getSize(buffer), f);
    fclose(f);

    sfBuffer_destroy(buffer);

    return 0;
}
```

**Review by eXpl0it3r — APPROVED** [review]

**Comment by eXpl0it3r** [comment]
@oprypin any comment from your side? 🙂

**Comment by oprypin** [comment]
It's good.
One thing is: putting `sfBuffer` into `graphics` module is strange because then we can't add a `sfBuffer` into `system` or `window` if we need it later. But it may be also strange to put it into `system` right away.

**Comment by eXpl0it3r** [comment]
Hmmm I put it off as "we can always move it later", but then we'd be breaking code. While I don't really hold myself to the same non-breaking standard for CSFML as for SFML, I think it might after all be beneficial to reduce it where possible.

So I think, it does after all matter and we should move `sfBuffer` to the system module. So it could be reused e.g. in the audio module in the future, without "breaking" the API.

**Comment by ChrisThrasher** [comment]
There are no other `saveToMemory` functions in SFML and if we added more they'd be added in SFML 3 and thus CSFML 3 so we can get away with leaving it where it is if we want. I'm fine moving it to System though. Either is fine with me.

**Comment by oprypin** [comment]
Hm but wait, if `sfBuffer` is first in `graphics` but then is moved to `system`, can't it just be re-exposed in `graphics` and then not break anything?

**Comment by oprypin** [comment]
But anyway probably `system` is better.  Sorry to cause you work :sweat_smile:

**Comment by eXpl0it3r** [comment]
> Hm but wait, if `sfBuffer` is first in `graphics` but then is moved to `system`, can't it just be re-exposed in `graphics` and then not break anything?

It could, but I wouldn't want such a workaround, just to avoid some potential breaking code. No need to add technical debt.

**Comment by kimci86** [comment]
I moved sfBuffer into csfml-system.

**Review by oprypin — APPROVED** [review]

**Review by eXpl0it3r — APPROVED** [review]
Thank you ❤️

**Pull request merged**